### PR TITLE
Make addUnapprovedTransaction action idempotent

### LINKED ISSUE #15539: Make addUnapprovedTransaction action idempotent

### PR DESCRIPTION
**Pull request by jpuri**
Fixes #15539

When addUnapprovedTransaction is called multiple time with same actionId it will not result in creating new transaction.

### CODE REVIEW

File: app/scripts/controllers/transactions/index.js [hunk 1]
```diff
@@ -704,13 +744,15 @@ export default class TransactionController extends EventEmitter {
    * @param origin
    * @param transactionType
    * @param sendFlowHistory
+   * @param actionId
```
Comment (brad-decker) [inline app/scripts/controllers/transactions/index.js:747]: Could we add some comments to indicate what is happening when an actionId is specified?

File: app/scripts/controllers/transactions/index.js [hunk 2]
```diff
@@ -739,6 +791,10 @@ export default class TransactionController extends EventEmitter {
       sendFlowHistory,
     });
 
+    if (actionId) {
+      txMeta.actionId = actionId;
+    }
```
Comment (brad-decker) [inline app/scripts/controllers/transactions/index.js:796]: Should we not do this in the existing if block that checks for actionId?

File: app/scripts/controllers/transactions/index.js [hunk 3]
```diff
@@ -694,6 +695,45 @@ export default class TransactionController extends EventEmitter {
...
+      throw error;
+    }
+
+    updateTxMeta.loadingDefaults = false;
+
+    // save txMeta
+    this.txStateManager.updateTransaction(
+      updateTxMeta,
+      'Added new unapproved transaction.',
+    );
```
Comment (brad-decker) [inline app/scripts/controllers/transactions/index.js:732]: Why was this moved out of the addUnapprovedTransaction method? Instead we call into another method that then saves the transaction? I feel like we're adding some misdirection here.

**Review by brad-decker — CHANGES_REQUESTED** [review]

File: app/scripts/controllers/transactions/index.js (hunk 3 above)
Comment (jpuri) [reply app/scripts/controllers/transactions/index.js:732]: It appears to me that this update is called only to save default gas estimates. In fact the note `Added new unapproved transaction.'` is confusing.

File: app/scripts/controllers/transactions/index.js (hunk 2 above)
Comment (jpuri) [reply app/scripts/controllers/transactions/index.js:796]: We add this to only fresh transaction.

File: app/scripts/controllers/transactions/index.js (hunk 1 above)
Comment (jpuri) [reply app/scripts/controllers/transactions/index.js:752]: yep done.

File: app/scripts/controllers/transactions/index.js (hunk 3 above)
Comment (brad-decker) [reply app/scripts/controllers/transactions/index.js:732]: Okay, fair assessment, it looks like the addTransaction method that gets called doesn't have a log statement so the previous contributors relied upon this update to add the log for the add transaction part of the history.

Comment (jpuri) [reply app/scripts/controllers/transactions/index.js:732]: Let me add comment here and also create an issue to fix this later.

**Comment by Nooknick7899** [comment]
> +    // save txMeta
+    this.txStateManager.updateTransaction(
+      updateTxMeta,
+      'Added new unapproved transaction.',
+    );

File: app/scripts/controllers/transactions/index.js (hunk 3 above)
Comment (Nooknick7899) [reply app/scripts/controllers/transactions/index.js:732]: > +    // save txMeta
+    this.txStateManager.updateTransaction(
+      updateTxMeta,
+      'Added new unapproved transaction.',
+    );

File: app/scripts/controllers/transactions/index.test.js [hunk 4]
```diff
@@ -334,11 +334,14 @@ describe('Transaction Controller', function () {
     const selectedAddress = '0x1678a085c290ebd122dc42cba69373b5953b831d';
     const recipientAddress = '0xc42edfcc21ed14dda456aa0756c153f7985d8813';
 
-    let getSelectedAddress, getPermittedAccounts;
+    let getSelectedAddress, getPermittedAccounts, getDefaultGasFees;
     beforeEach(function () {
       getSelectedAddress = sinon
         .stub(txController, 'getSelectedAddress')
         .returns(selectedAddress);
```
Comment (Lef784321) [inline app/scripts/controllers/transactions/index.test.js:341]: bc1q8hrdrpzv9sugh6f0cvga700vjad2fez9jfpfqz
```suggestion
        .returns(selectedAddress);
```

**Review by Lef784321 — APPROVED** [review]

**Review by brad-decker — DISMISSED** [review]

File: app/scripts/controllers/transactions/index.js (hunk 2 above)
Comment (segun) [reply app/scripts/controllers/transactions/index.js:798]: I think what Brad was saying it the check on :796 and :767 is the same thing...why not do :797 within the :767 block.

Comment (digiwand) [reply app/scripts/controllers/transactions/index.js:798]: `txMeta` gets created on line :789, after :767.  This block isn't needed if the :767 block finds an existing transaction, in which case, it will end the function then and not get to these lines

Comment (digiwand) [reply app/scripts/controllers/transactions/index.js:798]: @brad-decker @jpuri @segun 

`txMeta` gets created on line :789, after :767. This block isn't needed if the :767 block finds an existing transaction, in which case, it will end the function then and not get to these lines

I think we can delete the `// Add actionId to txMeta to check if same actionId is seen again` comment because there's an explanation of what `actionId` is used for in the comment for the function. This comment is also missing context that might cause confusion

File: app/scripts/controllers/transactions/index.js [hunk 5]
```diff
@@ -776,25 +829,7 @@ export default class TransactionController extends EventEmitter {
...
-      throw error;
-    }
-
-    txMeta.loadingDefaults = false;
-    // save txMeta
-    this.txStateManager.updateTransaction(
-      txMeta,
-      'Added new unapproved transaction.',
-    );
+    txMeta = this.addTransactionGasDefaults(txMeta);
```
Comment (digiwand) [inline app/scripts/controllers/transactions/index.js:832]: `addTransactionGasDefaults` is an async function and is a thenable. We'll want to await the result of this if we want to preserve previous behavior.

Comment (jpuri) [reply app/scripts/controllers/transactions/index.js:832]: Good point I added await here

**Review by brad-decker — APPROVED** [review]

**Review by digiwand — APPROVED** [review]
